README
== Welcome to ANTFARM

ANTFARM is a passive network mapping application that utilizes
output from existing network examination tools to populate its
OSI-modeled database.  This data can then be used to form a
'picture' of the network being analyzed.  This documentation is
consistent with version 0.3.0 of the ANTFARM software.

== More About ANTFARM

ANTFARM is a data fusion tool that does not directly interact
with the network.  The analyst can use a variety of passive or
active data gathering techniques, the outputs of which are
loaded into ANTFARM and incorporated into the network map.  Data
gathering can be limited to completely passive techniques
when minimizing the risk of disrupting the operational network
is a concern.

== Disclaimer

While the ANTFARM tool itself is completely passive (it does not
have any built-in means of gathering data directly from devices
or networks), network admin tools that users of ANTFARM may choose
to gather data with may or may not be passive.  The authors of
ANTFARM hold no responsibility in how users decide to gather data
they wish to feed into ANTFARM.

== Getting Started

As of now, the ANTFARM application is a command-line tool written
to be used in a Linux environment (hopefully someone will want to
make a Windows verson... :).  The following can be done to get up
and running with ANTFARM.

=== Installation

ANTFARM is packaged as a Ruby Gem, so the easiest way to install
it is to run 'gem install ANTFARM' from the command line.

The default database for ANTFARM is SQLite3, so at a minimum you will
need libsqlite3-dev installed.  The PostgreSQL database is also
supported (good for concurrent access), which requires the development
files for the PostgreSQL server to be installed, as well as the server
and client software and the postgres Ruby Gem.

=== Initialization

ANTFARM can be ran purely from within the Gem environment, but users
will not have the ability to (easily) save copies of the database and
use custom input/output scripts.  However, if the user executes

  antfarm db --initialize

from the command-line, a <tt>.antfarm</tt> directory will be created in the 
user's home directory.  Within this directory there will exist a <tt>config</tt>, 
<tt>db</tt>, <tt>log</tt>, and <tt>scripts</tt> directory.  The databases
and log files created by ANTFARM will reside here, and any custom scripts
written by the user can be placed here to be made available to the application.

Last, you need to create an actual database to store the data in.

  antfarm db --migrate

The above command will work out of the box if the SQLite3 database is used.
If you intend to use the PostgreSQL database, you must first create a database
in PostgreSQL.  The name of the database should match the name of the environment
you wish to use it in.

  su - postgres
  createuser <username> (you will only have to do this once)
  createdb antfarm (or whatever environment name you want)

Now you're ready to start parsing files and filling the database!

== Description of Contents

Add stuff here...

== Description of Defaults

Currently, the ANTFARM application is hard-coded to use an sqlite3 database.  
The name given for the ANTFARM environment (which defaults to 'antfarm') is 
the name used for the database file and the log file.

As of now, the application defaults to the __antfarm__ environment and to a
log level of __warn__.  These can be changed on a per-execution basis by passing
a <tt>-e</tt> and/or a <tt>-l</tt> option to the antfarm command.  These can
also be changed in the <tt>HOME_DIR/.antfarm/config/defaults.yml</tt> file so
you do not have to specify it on the command line every time.  This is probably
also the __safer__ way to go... if you forget one time to pass the <tt>-e</tt>
argument when using the non-default environment you could corrupt some data!

You can also tell the ANTFARM application to use the PostgreSQL database for
a particular environment by adding something similar to the following to your
defaults.yml file:

 <env name>:
   adapter: postgresql

For example, the following will tell ANTFARM to use the PostgreSQL database for
the default 'antfarm' environment:

 antfarm:
   adapter: postgresql

== Example Usage

Scripts are used to input data into the ANTFARM database.  Multiple scripts already
exist in the tool for parsing configuration files, traffic sniffs, etc.  To see
the scripts that currently exist, type

  antfarm help

Every command within ANTFARM also has some help information associated with it, so
typing something along the lines of

  antfarm help <command>

should be useful.

Below is a __very__ simple example of a script to read in network segments.  The 
script assumes it's parsing a file with one network defined per line.

  begin
    list = File.open(file)
  rescue Errno::ENOENT
    puts "The file '#{file}' does not exist"
    exit
  end

  list.each do |network|
    IpNetwork.create :address => network.strip
  end

The above code is acutally what's present in the <tt>load-network</tt> script.  To
run this script, you would type

  antfarm load-network <data_file>

Most built-in scripts have help text associated with them, so typing

  antfarm help <script_name>

will display some useful information.

Feel free to read the source code for the built-in scripts, which are located in the
<tt>ANTFARM_ROOT/lib/scripts/</tt> directory.  These will most likely be good examples
for learning how to create custom scripts.  You may add custom scripts to
<tt>HOME_DIR/.antfarm/scripts/</tt> directory and they will be available in the
ANTFARM application.

== Database Management

Obviously, there will be some times when you might need to manually edit some of the data
in the ANTFARM database.  You are more than welcome to access the database you are using
(which will be in <tt>HOME_DIR/.antfarm/db</tt> if you initialized your environment) with
the sqlite3 application.  However, for your convenience, a Rails application has been
included that utilizes the ActiveScaffold plugin (http://www.activescaffold.com) to allow
you to easily view and edit the database manually.  Simply type the following command:

  antfarm rails

and browse to http://localhost:3000/ using your favorite internet browser.

== Displaying Results

Currently, ANTFARM does not have any built-in ways of graphically viewing a network
(good spot for volunteers! :).  However, a Java application utilizing the open source 
Java Prefuse package (http://prefuse.sourceforge.net) has been created and included with
ANTFARM.  In order to use this application, the Java 6 Runtime Environment must be installed
on your computer.  A script has been provided to utilize this application.  It can be ran
by executing the following command:

  antfarm viz display-networks

This command will dump the data in the database to a temporary file (in the <tt>tmp</tt>
directory in your user space) then provide this file to the viz application.  Run it... it
looks cool! :)

If you look at <tt>HOME_DIR/.antfarm/config/colors.xml</tt>, you'll see the default file
this application uses to decide what colors to use for nodes and edges.  Modify this file
to get the colors you want.

== Homepage

http://antfarm.rubyforge.org

== RubyForge Project

http://rubyforge.org/projects/antfarm

== GitHub Project

http://github.com/btrichardson/antfarm

All code development will take place from GitHub.  The latest source code for each
ANTFARM release will be mirrored at the RubyForge project SCM as well.

== Legal Stuff

Copyright (2008) Sandia Corporation.
Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
the U.S. Government retains certain rights in this software.

This library is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at
your option) any later version.

This library is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
details.

You should have received a copy of the GNU Lesser General Public License
along with this library; if not, write to the Free Software Foundation, Inc.,
59 Temple Place, Suite 330, Boston, MA 02111-1307 USA


CHANGELOG
ANTFARM CHANGELOG

VERSION 0.5.0 (not released yet)
 * Added 'custom' string column to each table in database to be used however a developer of scripts needs to use it.
 * Added 'virtual' boolean column to ip_interface table to denote whether or not an IP interface is virtual (defaults to false)
 * Changed 'address' field in DNS table to 'ip_address' and added 'ethernet_address' field (this is why the minor version number had to change...)

VERSION 0.4.0
 * Added support for PostgreSQL database
 * Added ANTFARM console command, which initializes an IRB console with the ANTFARM environment
 * Added ANTFARM db --console option, which determines the database adapter being used and executes the appropriate database console command
 * Added ANTFARM cp-script command, which takes an existing ANTFARM core script and copies it to the user's custom script directory for use as a template
 * Updated Nmap XML parser script
 * Added visualization file for the traffic table
 * Modified the visualization file for networks to leave out traffic data

VERSION 0.3.0
 * Added Prefuse-based vizualization application and script to execute it with
 * Added defaults file so users can specify default environment and log level
 * Froze Rails to release 2.1.0
 * Updated ActiveScaffold plugin so it would work with Rails 2.1.0
 * Updated README documentation to be more specific on how to:
  * initialize the application
  * create a database
  * execute scripts
  * display the data
  * manually manipulate the data
  * set defaults
  * etc.
 * Updated README document with some disclaimers on ANTFARM being 'passive'
 * Created this CHANGELOG

VERSION 0.2.0
 * Initial version of project when open sourced.


lib/dbmanage.rb
# Copyright (2008) Sandia Corporation.
# Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
# the U.S. Government retains certain rights in this software.
#
# Original Author: Bryan T. Richardson, Sandia National Laboratories <btricha@sandia.gov>
#
# This library is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or (at
# your option) any later version.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA 

require 'fileutils'

module Antfarm

  # Extends SCParse::Command so it can be considered as a command.
  class DBManager < SCParse::Command
    def initialize
      # set the command name to 'db'
      super('db')

      @opts = OpenStruct.new
      @opts.clean = false
      @opts.initialize = false
      @opts.migrate = false
      @opts.remove = false
      @opts.reset = false
      @opts.console = false

      @options = OptionParser.new do |opts|
        opts.on('--initialize', "Initialize user's environment") do
          @opts.initialize = true
        end
        opts.on('--clean', "Clean the application's environment (remove all)") { @opts.clean = true }
        opts.on('--remove', "Remove existing database and log file for the given environment") do
          @opts.remove = true
        end
        opts.on('--migrate', "Migrate tables in database") do
          @opts.migrate = true
        end
        opts.on('--reset', "Reset tables in database and clear the log file for the given environment") { @opts.reset = true }
        opts.on('--console', "Start up an SQLite3 console using the database for the given environment") { @opts.console = true }
      end
    end

    def execute(args)
      super(args)

      if @opts.initialize
        init
      elsif @opts.remove
        db_remove
      elsif @opts.clean
        db_clean
      elsif @opts.migrate
        db_migrate
      elsif @opts.reset
        db_reset
      elsif @opts.console
        db_console
      end
    end

    #######
    private
    #######

    # Initializes a suitable directory structure for the user and
    # copies a default colors.xml file to the user's config directory
    def init
      FileUtils.makedirs("#{ENV['HOME']}/.antfarm/config")
      FileUtils.makedirs("#{ENV['HOME']}/.antfarm/db")
      FileUtils.makedirs("#{ENV['HOME']}/.antfarm/log")
      FileUtils.makedirs("#{ENV['HOME']}/.antfarm/scripts")
      FileUtils.makedirs("#{ENV['HOME']}/.antfarm/tmp")
      `cp #{ANTFARM_ROOT}/templates/defaults_template_file.yml #{ENV['HOME']}/.antfarm/config/defaults.yml`
      `cp #{ANTFARM_ROOT}/templates/colors_template_file.xml #{ENV['HOME']}/.antfarm/config/colors.xml`
    end

    # Removes the database based on the ANTFARM environment given
    def db_remove
      `rm #{Antfarm.log_file_to_use}` if File.exists?(Antfarm.log_file_to_use)
      `rm #{Antfarm.db_file_to_use}` if File.exists?(Antfarm.db_file_to_use)
    end

    # Removes any database, schema, and log files found
    def db_clean
      Find.find(Antfarm.db_dir_to_use) do |path|
        if File.basename(path) == 'migrate'
          Find.prune
        else
          `rm #{path}` unless File.basename(path) == 'db' || File.basename(path) == 'schema.rb'
        end
      end

      Find.find(Antfarm.log_dir_to_use) do |path|
        `rm #{path}` unless File.basename(path) == 'log'
      end

      Find.find(Antfarm.tmp_dir_to_use) do |path|
        `rm #{path}` unless File.basename(path) == 'tmp'
      end
    end

    # Creates a new database and schema file.  The location of the newly created
    # database file is set in the initializer class via the ActiveRecord 
    # configuration hash, and is based on the current ANTFARM environment.
    def db_migrate
      if File.exists?(File.expand_path("#{ANTFARM_ROOT}/db/schema.rb"))
        begin
          load(File.expand_path("#{ANTFARM_ROOT}/db/schema.rb"))
        rescue PGError
          puts "Looks like you are using the PostgreSQL database and you haven't yet created a database for this environment."
          puts "Please execute 'psql #{ANTFARM_ENV}' to create the database, then try running 'antfarm db --migrate' again."
        end
      else
        puts "A schema file did not exist. Running migrations instead."
        begin
          ActiveRecord::Migration.verbose = true
          ActiveRecord::Migrator.migrate(ANTFARM_ROOT + "/db/migrate/", nil)
          db_schema_dump if ActiveRecord::Base.schema_format == :ruby
        rescue PGError
          puts "Looks like you are using the PostgreSQL database and you haven't yet created a database for this environment."
          puts "Please execute 'psql #{ANTFARM_ENV}' to create the database, then try running 'antfarm db --migrate' again."
        end
      end
    end

    def db_reset
      db_remove
      db_migrate
    end

    def db_console
      if (defined? USER_DIR) && File.exists?("#{USER_DIR}/config/defaults.yml")
        config = YAML::load(IO.read("#{USER_DIR}/config/defaults.yml"))
      end
      puts "Loading #{ANTFARM_ENV} environment"
      if config && config[ANTFARM_ENV] && config[ANTFARM_ENV]['adapter'] == 'postgresql'
        exec "psql #{ANTFARM_ENV}"
      else
        exec "sqlite3 #{Antfarm.db_file_to_use}"
      end
    end

    def db_schema_dump
      require 'active_record/schema_dumper'

      File.open(File.expand_path("#{ANTFARM_ROOT}/db/schema.rb"), "w") do |file|
        ActiveRecord::SchemaDumper.dump(ActiveRecord::Base.connection, file)
      end
    end
  end

end
